project/CelebornBuild.scala

```diff
@@ -62,34 +62,46 @@ object Dependencies {
   val protocVersion = "3.19.2"
   val protoVersion = "3.19.2"
   
-  val commonsCrypto = "org.apache.commons" % "commons-crypto" % commonsCryptoVersion
+  val commonsCrypto = "org.apache.commons" % "commons-crypto" % commonsCryptoVersion excludeAll(
+    ExclusionRule("net.java.dev.jna", "jna"))
   val commonsIo = "commons-io" % "commons-io" % commonsIoVersion
   val commonsLang3 = "org.apache.commons" % "commons-lang3" % commonsLang3Version
   val findbugsJsr305 = "com.google.code.findbugs" % "jsr305" % findbugsVersion
   val guava = "com.google.guava" % "guava" % guavaVersion
-  val hadoopClientApi = "org.apache.hadoop" % "hadoop-client-api" % hadoopVersion
-  val hadoopClientRuntime = "org.apache.hadoop" % "hadoop-client-runtime" % hadoopVersion
-  val ioDropwizardMetricsCore = "io.dropwizard.metrics" % "metrics-core" % metricsVersion
-  val ioDropwizardMetricsGraphite = "io.dropwizard.metrics" % "metrics-graphite" % metricsVersion
-  val ioDropwizardMetricsJvm = "io.dropwizard.metrics" % "metrics-jvm" % metricsVersion
-  val ioNetty = "io.netty" % "netty-all" % nettyVersion
+  val hadoopClientApi = "org.apache.hadoop" % "hadoop-client-api" % hadoopVersion excludeAll(
+    ExclusionRule("org.slf4j", "slf4j-api"))
```

Comment:
does sbt support globally pin dep's version?

Reply:
Certainly, the [SBT](https://www.scala-sbt.org/1.x/docs/Library-Management.html#Forcing+a+revision+%28Not+recommended%29) does allow for the specification of a forced dependency revision, e.g.

```
libraryDependencies ++= Seq(
  "log4j" % "log4j" % "1.2.14" force()
)
```

However, the docs also point out it is advised against, as indicated by the following considerations:

> 1. Forcing can create logical inconsistencies so it’s no longer recommended.
> 2. this is an Ivy-only feature and cannot be included in a published pom.xml.

Reply:
an other approach, [Disable Transitivity](https://www.scala-sbt.org/1.x/docs/Library-Management.html#Disable+Transitivity) 

> By default, these declarations fetch all project dependencies, transitively. In some instances, you may find that the dependencies listed for a project aren’t necessary for it to build. Projects using the Felix OSGI framework, for instance, only explicitly require its main jar to compile and run. Avoid fetching artifact dependencies with either intransitive() or notTransitive(), as in this example:
> 
> ```
> libraryDependencies += "org.apache.felix" % "org.apache.felix.framework" % "1.8.0" intransitive()
> ```

Reply:
Perform local testing. the `intransitive()` cannot achieve this functionality.
```
[warn] Found intransitive dependency (org.apache.hadoop:hadoop-client-api:3.2.4 intransitive) while publishMavenStyle is true, but Maven repositories
[warn]   do not support intransitive dependencies. Use exclusions instead so transitive dependencies
[warn]   will be correctly excluded in dependent projects.
```

Reply:
Let's use `force()` way.

> 1. Forcing can create logical inconsistencies so it’s no longer recommended.

We know what we are doing, IMO, rules-based dependency derivation is more unpredictable.

> 2. this is an Ivy-only feature and cannot be included in a published pom.xml.

It does not affect our use case, since we only publish client-shaded modules, without transitive deps.

Reply:
Test locally with commit https://github.com/apache/incubator-celeborn/pull/1831/commits/32f7ce052763a12c8f9f33515e26143f498ea297, The use of `force()` ineffective.

Reply:
I've also come across an alternative method for setting up dependencyOverrides described here: [Overriding a Version](https://www.scala-sbt.org/1.x/docs/Library-Management.html#Overriding+a+version).

Moreover, during my search, I observed that prominent projects employing SBT as their build tool – like [akka](https://github.com/akka/akka/blob/d7b17e9dec52cd27cd17192dda58598d229e3985/project/Dependencies.scala#L96), [play](https://github.com/playframework/playframework/blob/d07e378e88471a4ec3613ebc842d4581143b047b/project/Dependencies.scala#L285), and [delta](https://github.com/delta-io/delta/blob/23826a3b25f39cc1097d3b6174f009da496028b7/build.sbt#L434) – commonly adopt the excludeAll approach to exclude transitive dependencies.

Reply:
Explicitly Exclude from all deps looks ugly. `dependencyOverrides` seems as clean as `force()`? if yes, let's try this way.

Reply:
addressed in https://github.com/apache/incubator-celeborn/pull/1831/commits/269b9cedb28be15a0d60843fe3b1c293af6d699a, it looks more clearly now.

Reply:
```
$ ./build/sbt -Pspark-3.3 celeborn-client-spark-3/dependencyTree | grep slf4j-api
[info]   | | | +-org.slf4j:slf4j-api:1.7.36
[info]   | | | | +-org.slf4j:slf4j-api:1.7.36
[info]   | | | +-org.slf4j:slf4j-api:1.7.36
[info]   | | | | +-org.slf4j:slf4j-api:1.7.36
...
```

Reply:
nice

Reply:
BTW, `dependencyOverrides` seems does not support pattern, e.g. `io.netty:*`, right?